### ISSUE TITLE
Add a factory for routing

### ISSUE THREAD

**Issue by NervosaX** [cerebral/cerebral-addons#13]
So currently I have a factory of my own that navigates to a page. It can generally be called like so:

`navigateTo("path/to/route")`

However there are more often times that I need to route to a page with an id, based on an input or the current route in some way. Calling `navigateTo("path/to/route/1")` is only possible statically, or wrapped inside another action specifically created for routing to that one page. 

It would be nice to have a syntax either especially for routing, or for creating a path dynamically based on input or state.

A possible example:

```javascript
[
  // outputs id: 1
  createSomething, {
    // Navigates to "/path/to/1"
    success: [navigateTo("/path/to/{input:/id}")]
    error: []
  }
]
```

**Comment by Guria** [comment]
@NervosaX why do you even need an `navigateTo` some url directly. It sticks you to some url. The great part of `cerebral` that you can avoid tying to source params. I am sure that `cerebral` app should always be able to work without router module at all. 
Please provide more details on your case.

**Comment by NervosaX** [comment]
@Guria Well, I want it to call the signal I have on routing to the detail page in the same way as if I load the page directly - and I can't really have a signal call the other signal. I could share a chain, but that feels like too much upkeep.

So the /appointment/create page has my form, you press submit, it then navigates and calls my appointment detail page, which fetches all the information about that appointment. If I go to that detail page at any other time, it just calls the detail page's signal on page open.

**Comment by Guria** [comment]
@NervosaX My strong recommendation: "do not attach any business logic to routable signals". So workflow would be like:
* routable signal mutates only state to indicate that details page should be shown.
* application mounts appropriate component
* details component triggers a signal in `componentDidMount` lifecycle callback, where you can fetch any data you need.

Does it makes sense?

**Comment by NervosaX** [comment]
Yeah, that sounds reasonable. I'll try it out.

**Comment by Guria** [comment]
@NervosaX [Followed your demo](https://github.com/NervosaX/cerebral-issue/pull/1) and got your point.

*Conclusion*
We should be able redirect to signals not tied to plain urls, so we don't write things like:
```js
({input, services}) => {
  services.router.redirect('/' + input.id)
}
```

*Proposal*
we can extend router redirect service with support of `services.router.redirect('details', {id: input.id})`
and then make addon (maybe also a part of router itself) to call: `redirect('details', {id: 'input:/id'})`

Does it fit your need?

**Comment by NervosaX** [comment]
Yes, I believe it would

**Comment by Guria** [comment]
redirectToSignal service is on the way
https://github.com/cerebral/cerebral-module-router/pull/65

**Comment by NervosaX** [comment]
@Guria This works really well, but it would be nice to have this as some sort of addon as well. I find myself having to create lots of one liner actions to use the redirectToSignal service.

Perhaps something like the following:

For route: /path/:firstId/to/:secondId

```javascript
navigateToSignal("path.to.viewOpened", {
  firstId: "input:/firstId",
  secondId: "input:/secondId"
});
```

**Comment by Guria** [comment]
@NervosaX yeap. I need @garth advice on how to implement it and which package should include it.

PS. The more a thinking then more I making sure that router should not deal with `redirect*` at all. I still think that app must able to work without router at all. The only things router should enable is multiple entry points to app. But it lead us to requirement of calling signals from signals again :(

I need @christianalfoni opinion here

**Comment by christianalfoni** [comment]
Just to summarize some approaches with the example above.

Signal path: "path.to.viewOpened"
Payload: input.firstId, input.secondId

1. Bring in the chain of path.to.viewOpened, and compose.
Good: We compose and it runs as one signal, very easy to read and understand
Problem: Router does not update

2. We actually trigger path.to.viewOpened signal
Good: The router updates
Problem: We fire off a signal inside a signal, which is considered a bad practice. Makes code harder to read and reason about. You can technically add more actions after the redirect, which makes it very hard to reason about what is happening.

The challenge here is that we are mixing routing redirects with signal redirects. I have been thinking a bit and with an application without a router you do not need "signal redirects", you just compose signals together. With the router we get into a different situation because of the url that needs to update.

So with the example above, not having arouter:

```js
[
  // outputs id: 1
  createSomething, {
    // Open id 1
    success: [...openSomething]
    error: []
  }
]
```

Composes, reads well and runs as one signal. In my opinion, that is optimal. 

But now we want the router to update itself. So from the perspective of keeping the benefits I think we should do something like:

```js
[
  // outputs id: 1
  createSomething, {
    // Open id 1
    success: route(openSomething)
    error: []
  }
]
```

The factory will create an action that can reference *openSomething* to figure out which route it belongs to, but it is also just returned. Basically it looks like this:

```js
[
  // outputs id: 1
  createSomething, {
    // Open id 1
    success: [routeAction, ...doSomething]
    error: []
  }
]
```

So what happens inside the factory is something like this:

```js
function route(chain) {

  function routeAction({input, services}) {
    // Iterates routes to find chain and updates url
    services.router.updateRouteByChain(chain, input);
  }

  return [
    routeAction,
    ...chain
  ];

}
```

If there are specific keys you want to use as payload we could allow choosing them with an array for example:

```js
[
  // outputs id: 1
  createSomething, {
    // Open id 1
    success: route(openSomething, ['firstId', 'secondId'])
    error: []
  }
]
```

So to summarize. Embracing composition makes signals more readable, more flexible and it Cerebral will also understand them better. With the above code you could even do:

```js
[
  // outputs id: 1
  createSomething, {
    // Open id 1
    success: [someSuccess, ...route(openSomething, ['firstId', 'secondId']), afterRoute]
    error: []
  }
]
```

Remember, the URL is just a state in your app... "just an input with some text" ;-)

**Comment by Guria** [comment]
@christianalfoni great. there is something in this. I will make another round of thoughts.

**Comment by bfitch** [comment]
I have a use case that is very similar, I believe. Consider a simple todo app with routes: 
```js
 router: Router({
    '/': 'rootRouted',
    '/todos/:id': 'showTodo'
  })
```

In the TodoList component, you can click a todo to navigate to its details page:
```js
class TodoList extends React.Component {
  todoTitleClicked(id) {
    this.props.signals.showTodo({id: id.toString()});
  }

  renderTodos() {
    return (this.props.todos).map(todo => {
      return (
        <ul>
          <li onClick={()=>{this.todoTitleClicked(todo.id)}}>{todo.title}</li>
          <li>{todo.user_id}</li>
        </ul>
      );
    })
  }
...
```
In the `showTodo` signal, I'm trying to use a module that exposes a REST/CRUD interface for fetching data (either locally or from the server). In the example below, `find()` is a chain factory that implements all the actions for checking the cache and making ajax requests, etc. 
```js
 showTodo: [
      ...find('todos', {id: input.id})({
        success: [
          ({input, state}) => {
              state.set('activeTodo', input.todos);
          },
          ({state}) => {
            state.set('activePage', 'ShowTodo');
          }
        ]
      })
  ]
``` 
The problem I'm running into is similar to @NervosaX, I believe. `find()` needs access to `input` to fetch its data and it can't be hardcoded, since it needs to handle whatever is passed into the signal: id, uuid, lastName, etc.

I originally implemented `find()` as a service, but realized it was much more flexible to *return* the cached or fetched data using `outputs` to a chain (`{success: []}`, for example sets the active todo that was fetched). Before, the service merely trigged some signals in the module and set data.

My first instinct was to create an action factory. But, I couldn't figure out how the action could return a chain. Now `find()` simply returns a chain but it doesn't have access to `input`...so I'm going in circles. :stuck_out_tongue_winking_eye: 

I'm not sure if addons can help with this or if there's another approach, but seems very similar to the issue here. Thanks for any suggestions. :)

**Comment by christianalfoni** [comment]
This is why we have put `cerebral-url-scheme-compiler` in its own project :-)

Then you can:

```js
 showTodo: [
      ...find('todos', 'input:/id')({
        success: [
          ({input, state}) => {
              state.set('activeTodo', input.todos);
          },
          ({state}) => {
            state.set('activePage', 'ShowTodo');
          }
        ]
      })
  ]
```

**Comment by christianalfoni** [comment]
Here is an example:

```js
var getCompiler = require('cerebral-url-scheme-compiler/get');

function entityHas(entityType, uuidPath, keys) {

  function action(args) {

    var output = args.output;
    var module = args.modules['cerebral-module-entities'];
    var getValue = getCompiler(uuidPath);

...
```

**Comment by christianalfoni** [comment]
Cleaning up, reopen if not solved :)

**Comment by NervosaX** [comment]
@christianalfoni @Guria Don't think the problem here has been solved (without having to create an action for each navigation)
